### PR TITLE
[ROX-12625 + ROX-13032] : Add GetCollectionCount and UpdateCollection endpoints and  services

### PR DESCRIPTION
**Pull request by charmik-redhat**
## Checklist
- [ ] Investigated and inspected CI test results
- [X] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

Added unit tests

In addition to reviewing your code, reviewers **must** also review your testing
instructions and make sure they are sufficient.

### CODE REVIEW

File: central/resourcecollection/service/service_impl.go
Comment (charmik-redhat) [reply central/resourcecollection/service/service_impl.go]: typo, fixed it

File: central/resourcecollection/service/service_impl.go
Comment (charmik-redhat) [reply central/resourcecollection/service/service_impl.go:161]: I mean the loop at the end of `collectionRequestToCollection` func.

```
if len(request.GetEmbeddedCollectionIds()) > 0 {
		embeddedCollections := make([]*storage.ResourceCollection_EmbeddedResourceCollection, 0, len(request.GetEmbeddedCollectionIds()))
		for _, id := range request.GetEmbeddedCollectionIds() {
			embeddedCollections = append(embeddedCollections, &storage.ResourceCollection_EmbeddedResourceCollection{Id: id})
		}
		collection.EmbeddedCollections = embeddedCollections
	}
```

File: central/resourcecollection/service/service_impl.go
Comment (md2119) [inline central/resourcecollection/service/service_impl.go:107]: The error message it cryptic. how about `errors.Errorf("%s env var is not enabled", {feature_flag}.EnvVar())`

File: central/resourcecollection/service/service_impl.go
Comment (md2119) [inline central/resourcecollection/service/service_impl.go:171]: You need to fetch the stored collection object and reconcile the updated fields with it i.e. the `createdAt` and `createdBy`. Update the `UpdateCollection` func and add a unit test.

**Review by md2119 — CHANGES_REQUESTED** [review]

File: central/resourcecollection/service/service_impl.go
Comment (charmik-redhat) [reply central/resourcecollection/service/service_impl.go:171]: Good catch, made that change on `UpdateCollection` implementation on datastore instead as datastore fetches the storedCollection anyways.

**Review by md2119 — APPROVED** [review]
The error handling in some scenarios is crude. The error messages do not provide enough context. For example, if embedded collections are not present, the message returned is `{id} is unknown`. Note that users are not working with IDs in the UI therefore, providing info that cannot even be found in the UI does not help. A message like `attached collection {collection_name} not found.` pinpoints users to the cause immediately.

Additionally, above scenario is not unit-tested.

I am not blocking this PR because there are multiple such instances however address this issue as soon as possible. Create a TODO JIRA.

File: central/resourcecollection/service/service_impl.go
Comment (charmik-redhat) [reply central/resourcecollection/service/service_impl.go:169]: `collectionRequestToCollection` takes in a common interface `collectionRequest` for CreateCollectionRequest and UpdateCollectionRequest. Since Id is not a field of CreateCollectionRequest, that method can't be added to `collectionRequest` interface. I would revert to original idea of passing in an `idGetter` func argument for updates.

**Review by keyallis — APPROVED** [review]
Noticed we're checking for resource selectors to not be empty, not sure if that's something we actually need to be checking.

Make sure to update the error comment before merging

File: central/resourcecollection/service/service_impl.go
Comment (md2119) [reply central/resourcecollection/service/service_impl.go:169]: Currently, the `collectionRequest` does not represent any collection request, it represents one type of request quite contradictory to the interface.

There are a couple of ways you can resolve, I will list a few:
**1. By changing the interface**
- Change the interface to:
```
type collectionRequest interface {
	GetID() string
	GetName() string
	GetDescription() string
	GetResourceSelectors() []*storage.ResourceSelector
	GetEmbeddedCollectionIds() []string
}
```
- Implement a wrapper for create request:
```
type createRequestWrapper struct {
	*v1.CreateCollectionRequest
}

func (*createRequestWrapper) GetID() string { return "" }

```
- Wrap the create request in the func call:
```
collection, err := collectionRequestToCollection(ctx, &createRequestWrapper{request}, true)
```
**2. Change the func signature** 
-  Change the function signature to receive `proto.Message` interface `collectionRequestToCollection(ctx context.Context, msg proto.Message)`. It removes dependency on `isCreate`
- Cast the incoming request:
```
if request, ok := msg.(*v1.CreateCollectionRequest); ok {
// do something
} else if request, ok := msg.(*v1.UpdateCollectionRequest); ok {
// do something
}
```

**Pull request merged**
